mava/distributions.py

```diff
@@ -91,6 +92,61 @@ def _parameter_properties(cls, dtype: Optional[Any], num_classes: Any = None) ->
         return td_properties
 
 
+class MaskedEpsGreedyDistribution(tfd.Categorical):
+    """
+    Computes an epsilon-greedy distribution for each action choice. There are two
+    components in the distribution:
+
+    1. A uniform component, where every action that is NOT masked out gets an even weighting.
+    2. A greedy component, where the action with the highest corresponding q-value that is
+    NOT masked out gets a probability of one.
+
+    Combining these two distributions per action choice in a ratio of eps:1-eps gives
+    the final distribution. This distribution can be sampled using mode() for a purely
+    greedy strategy, and sampled normally using sample() for an epsilon-greedy strategy.
+    """
+
+    def __init__(self, q_values: chex.Array, epsilon: float, mask: chex.Array):
+        self.q_values = q_values
+
+        # UNIFORM PART (eps %)
+        # generate uniform probabilities across all allowable actions at most granular level
+        masked_uniform_action_probs = mask.astype(int)
+        # get num avail actions to generate probabilities for choosing
+        n_available_actions = jnp.sum(masked_uniform_action_probs, axis=-1)[..., jnp.newaxis]
+        # divide with sum along axis to get uniform per action choice
+        masked_uniform_action_probs = masked_uniform_action_probs / n_available_actions
+
+        # GREEDY PART (1-eps %)
+        # set masked actions to value not chosen by argmax
+        masked_q_vals = jnp.where(
+            mask,
+            q_values,
+            jnp.finfo(jnp.float32).min,
+        )
+
+        # greedy argmax over action-value dim
+        greedy_actions = jnp.argmax(masked_q_vals, axis=-1)
+        # get one-hot so that shapes are equal again and ready to be made into a prob
+        greedy_actions = jax.nn.one_hot(greedy_actions, q_values.shape[-1])
+        # consistency check
+        chex.assert_equal_shape([greedy_actions, masked_q_vals, q_values])
+
+        mixed_eps_greedy_probs = (
+            epsilon * masked_uniform_action_probs + (1 - epsilon) * greedy_actions
+        )
+
+        super().__init__(probs=mixed_eps_greedy_probs)
+
+    def get_q_values(self) -> chex.Array:
+        return self.q_values
```

Comment:
In python I'd say this should usually be a property so we'd store `self._q_values` and then make this method:
```
    @property
    def q_values(self) -> chex.Array:
        return self._q_values
```

Although to be honest I think we can just leave this method out and people can just do `dist.q_values`

Reply:
Got it, I'll go for the simpler version of dist.q_values!